flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java

```diff
@@ -64,6 +64,8 @@
 
     private TimerGauge backPressuredTimeMsPerSecond = new TimerGauge();
 
+    private long totalWrittenBytes;
```

Comment:
can we use this variable to replace `numBytesOut` `Counter` and metric?

Reply:
Of course, we can rewrite the metric in such a way that instead of passing the counter it can call the method for receiving the number of bytes. But it definitely is not one line change and if we do so unfortunately semantic for `numBuffersOut` and `numBytesOut` will be different which is not tragic but anyway. 
So we can do this but I don't sure that it makes sense to do it here. Or do you think that the separate commit under this ticket is ok?

Reply:
If we decide to do that, I think it should be done under this ticket to avoid acquiring a technological debt (in a separate commit).

However keep in mind that I haven't thought deeply about this deduplication. If it adds a lot of complexity it might be actually cleaner to have them duplicated.

Reply:
I have thought about that. I think it is better to leave them duplicated since right now we have `TaskIOMetricGroup` which handles the metrics the same - `Counter` + `MeterView` but if we want to change it we need to do at least the following:

- We need to adapt `MeterView` to receive value directly from some entity rather than from `Counter`(or create a wrapper for counter). This is not a big problem but anyway, this should be done.
- We need to initialize one particular metric differently since we won't be able to create this metric in `TaskIOMetricGroup` constructor. This is a more serious problem because it can lead to confusion if we start to initialize metrics from one group differently in different places.
- We need to implement the same logic of `totalWrittenBytes` in other subclasess of `ResultPartition` since right now it is just a helper for calculation of metric in one specific implementation but if we decide to make it as metric we need to have it everywhere(at least in `SortMergeResultPartition`). Again, it is not a big problem but just an extra complication.

So right now I think that it doesn't make sense to change anything in `TaskIOMetricGroup` but in general, it is a good idea to think about our metrics. Because the metric which collects data directly from the entity by demand looks more flexible than the metric that passes the counter(or other metric-related things) to the entity. But this is not today's discussion.